src/plugins/scopes/Scope.js

```diff
@@ -0,0 +1,204 @@
+                if (varDepth === depth) {
+                    if (typeOrder[variable.type] < typeOrder[refVar.type]) {
+                        refVar._transferReferences(variable);
+                        removeVariableIfRequired(refVar);
+                    }
+                } else if (varDepth < depth) {
+                    refVar._references.delete(reference);
+                    variable._addReference(reference);
+                    reference._variable = variable;
+                    removeVariableIfRequired(refVar);
+                }
+            }
+        }
+
+        for (let childScope of this._childScopes) {
+            childScope._adjustReferencesOnVariableAdd(variable);
+        }
+    }
+
+    _addReference(referenceInfo: {node: Node, name: string, read: boolean, write: boolean}) {
+        let {name} = referenceInfo;
+        let reference = new Reference({scope: this, ...referenceInfo});
+        this._assignReference(reference, name);
+        let references = this._references.get(name);
+        if (references) {
+            references.push(reference);
+        } else {
+            this._references.set(name, [reference]);
+        }
+    }
+
+    _assignReference(reference: Reference, name: string) {
+        let currentScope = this;
+        do {
+            let variables = currentScope._variables.get(name);
+            if (variables) {
+                variables[0]._addReference(reference);
+                break;
+            }
+            if (!currentScope._parentScope) {
+                let globalVariable = new Variable({
+                    name, type: types.ImplicitGlobal, scope: currentScope
+                });
+                globalVariable._addReference(reference);
+                currentScope._addVariable(globalVariable);
+                break;
+            } else {
+                if (
+                    (
+                        (name === 'arguments' || name === 'this') &&
+                        currentScope._isFunctionScope &&
+                        !currentScope._isArrowFunctionScope &&
+                        !currentScope._isProgramScope
+                    ) ||
+                    (
+                        name === 'super' && currentScope._isClassScope
+                    )
+                ) {
+                    let builtInVariable = new Variable({
+                        name, type: types.BuiltIn, scope: currentScope
+                    });
+                    builtInVariable._addReference(reference);
+                    currentScope._addVariable(builtInVariable);
+                    break;
+                }
+                currentScope = currentScope._parentScope;
+            }
+        } while (true);
+    }
+
+    get node(): Node {
+        return this._node;
+    }
+
+    get parentScope(): ?Scope {
+        return this._parentScope;
+    }
+
+    get variables(): Variable[] {
+        return [].concat.apply([], Array.from(this._variables.values()));
```

Comment:
Oh why was this necessary rather than just `Array.from(this._variables.values()` which returns an array?